InteractionBase/src/vis/SelectionAtCursorItem.cpp

```diff
@@ -0,0 +1,165 @@
+{
+	currentNodes_ = arrange(selectableNodes);
+	target->scene()->addTopLevelItem(this);
+	mousePosition_ = target->scene()->lastMouseHoverPosition();
+	selectedEffect_ = new QGraphicsColorizeEffect();
+	if (hasTextField)
+	{
+		textField_ = new Visualization::Text(this, &style->nameField(), "Enter name here");
+		textField_->setEditable(true);
+		selectItem(textField_);
+	}
+}
+
+void SelectionAtCursorItem::initializeForms()
+{
+	auto nodeGrid = (new Visualization::DynamicGridFormElement())
+			->setSpacing(5, 5)
+			->setMajorAxis(Visualization::GridLayouter::ColumnMajor)
+			->setNodesGetter([](Visualization::Item* v)
+				{ auto self = static_cast<I*>(v);
+				  return self->currentNodes(); });
+	//First form: only the nodes
+	addForm(nodeGrid);
+	//Second form: the nodes and a text field
+	addForm((new Visualization::GridLayoutFormElement())
+			->setSpacing(5, 5)->setMargins(10)
+			->put(0, 0, item<Visualization::Text>(&I::textField_,
+									[](I* v) -> const Visualization::TextStyle* { return &v->style()->nameField(); }))
+									//[](I* v) { return v->style()->nameField(); }))
+			->put(0, 1, nodeGrid));
+}
+
+int SelectionAtCursorItem::determineForm()
+{
+	return hasTextField() ? 1 : 0;
+}
+
+void SelectionAtCursorItem::selectItem(Visualization::Item *item)
+{
+	if (focusedItem() && focusedItem() != textField())
+	{
+		focusedItem()->setGraphicsEffect(nullptr);
+		selectedEffect_ = new QGraphicsColorizeEffect();
+	}
+	if (item && item != textField())
+		item->setGraphicsEffect(selectedEffect_);
+	if (item == textField())
+	{
+		item->moveCursor();
+		item->correspondingSceneCursor<Visualization::TextCursor>()->selectAll();
+	}
+	else moveCursor(Visualization::Item::MoveOnPosition, QPoint(0, 0));
+	focusedItem_ = item;
+}
+
+void SelectionAtCursorItem::executeFocused()
+{
+	if (hasTextField() && focusedItem() == textField() && onText_)
+		onText_(textField()->text(), target_);
+	else if (focusedItem() && onSelect_)
+		onSelect_(focusedItem()->node(), target_);
+}
+
+QPoint SelectionAtCursorItem::indexOf(Model::Node *node) const
+{
+	for (int col = 0; col < currentNodes_.size(); col++)
+		for (int row = 0; row < currentNodes_[col].size(); row++)
+			if (currentNodes_[col][row] == node)
+				return QPoint(col, row);
+	return QPoint(-1, -1);
+}
+
+QPoint SelectionAtCursorItem::correctCoordinates(QPoint point) const
+{
+	QPoint result = point;
+	if (point.x() < 0)
+		result.setX(0);
+	if (point.x() >= currentNodes_.size())
+		result.setX(currentNodes_.size() - 1);
+	if (point.y() < 0)
+		result.setY(0);
+	if (point.y() >= currentNodes_[result.x()].size())
+		result.setY(currentNodes_[result.x()].size() - 1);
+	return result;
+}
+
+void SelectionAtCursorItem::updateGeometry(int availableWidth, int availableHeight)
+{
+	Super::updateGeometry(availableWidth, availableHeight);
+	setPos(mousePosition_.x() - widthInScene() / 2, mousePosition_.y() - heightInScene() / 2);
+	if (!focusedItem())
```

Comment:
I'm not sure why we need this. We should not automatically focus an item in updateGeometry. What you can do in determineChildren is call `setDefaultMoveCursorProxy(childItem);` to set what child item should be automatically selected **if** someone request an automatic focus. Which should happen from the outside, in the show method.